torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3663,7 +3663,16 @@ def add_expr(expr):
             # variables have to have their value range bounds adjusted as
             # well)
             s = sympy.Symbol(f"shape_{idx}", positive=True, integer=True)
-            offset = lower - 1
+
+            # Note:
+            #   Offset might be a fraction(e.g. aten.split.Tensor), but shapes are always integers.
+            #   Sympy might give unexepected results when comparing an integer with a non-integer
+            #   Therefore, we cast offset to int here.
+            #   For example:
+            #       shape_0 = sympy.Symbol("shape_0", positive=True, integer=True)
+            #       expr = sympy.Eq(shape_0 - 1/3, 4)
+            #       expr.xreplace({}) # False
+            offset = int(lower - 1)
```

Comment:
In lowering of aten.aplit, sympy expression `Eq(shape_0 - 1/3, 4)` is evaluated by xreplace. However, the result is False because shape_0 - 1/3 is fraction.

Reply:
This can be reproduced by torch:
```python
        @torch.compile()
        def f(x, sizes):
            return torch.split(x, sizes, -1)
        # split into equally sized chunks, 12 = 4 + 4 + 4
        r1, r2, r3 = f(torch.randn(2, 12), 4)
```
or by sympy
```python
In [9]: shape_0 = sympy.Symbol("shape_0", positive=True, integer=True)
In [10]: expr = sympy.Eq(shape_0 - 1/3, 4)]
In [11]: expr.xreplace({})
Out[11]: False
```

Reply:
@jansel @ezyang Please enlighten my if any better solutions. Should I use ceil or floor instead?

Reply:
Can we add a `sympy.Floor` to the shape calculation inside of the `torch.split` lowering?

It seems a bit odd to handle it here.

Reply:
> Can we add a `sympy.Floor` to the shape calculation inside of the `torch.split` lowering?
> 
> It seems a bit odd to handle it here.

I understand your concern because this change could possibly affects all ops. But IMO handling it here makes sense, bacause shape are expected to be integer. BTW, The only failed CI was unrelated to this change.

But I am new to inductor. If my thoughts is incorrect please tell me and I can handle it in `torch.split` lowering.

Reply:
> On vacation and only did a skim of the PR, but ValueRanges should never have non integral bounds for int variables, so I would figure out why this happened first

For this example:
```
        @torch.compile(dynamic=True) #set to True
        def f(x, sizes):
            return torch.split(x, sizes, -1)
        # split into equally sized chunks, 12 = 4 + 4 + 4
        r1, r2, r3 = f(torch.randn(2, 12), 4)
```
shape0 = `x.shape[1] , range=(lower=2, upper=92223372....)`
shape1 = `x.shape[1] // sizes`
Guard is `3 * shape1 <= shape0`, thus range of shape1 is `(lower=2/3, upper=....)`

Reply:
Ya we should end up with the more conservative bound 0 here

Reply:
I'm going to look into moving the clamping earlier

Reply:
Interestingly, the non-integer range gets introduced when we do this guard: `Eq(s2, s1 - 2*s2)`